### ISSUE TITLE
Flink Checkpoints not updating ReaderGroup positions in new Pravega artifacts

### ISSUE THREAD

**Issue by RaulGracia** [pravega/pravega#3076]
**Problem description**
In PR https://github.com/pravega/pravega-samples/pull/161, we are updating the Flink `StreamCut`s sample app to use the new `generateStreamCuts()` call. In a nutshell, the application relies in Flink's checkpointing mechanism to get the initial `StreamCut` to define a stream "slice", and now it uses the `generateStreamCuts()` call to get the end of a stream slice. Apparently, this sample app uncovered a potential bug: Flink checkpoints seem to do not complete, which yields that the `ReaderGroup` reading positions is not being updated. As a side effect, this makes the application unable to get a correct start from a stream slice.

To discard that problems come from the new version of the application, I also performed tests with the [existing working version](https://github.com/pravega/pravega-samples/tree/develop/flink-connector-examples/src/main/java/io/pravega/example/flink/streamcuts). This means that in these tests both Flink version (`1.6`) and the app code are the same as before, so the only components that have been updated are:
```
pravegaVersion=0.4.0-50.9a734d2-SNAPSHOT -> pravegaVersion=0.4.0-50.da91e55-SNAPSHOT
flinkConnectorVersion=0.4.0-100.d588197-SNAPSHOT -> flinkConnectorVersion=0.4.0-96.54ccab2-SNAPSHOT 
```

In this situation, the `StreamBookmarker` job shows the following output:
```
10:23:41,179 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Start bookmarking a stream slice at: examples11/sensor-events:0=0, 1=0, 2=0 for sensor 2.
10:23:49,683 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Start bookmarking a stream slice at: examples11/sensor-events:0=0, 1=0, 2=0 for sensor 1.
10:24:01,387 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Start bookmarking a stream slice at: examples11/sensor-events:0=0, 1=0, 2=0 for sensor 0.
10:24:14,604 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Initialize sensorStreamSlice end value to look for the next updated StreamCut: examples11/sensor-events:0=0, 1=0, 2=0 for sensor 2.
10:24:26,210 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Initialize sensorStreamSlice end value to look for the next updated StreamCut: examples11/sensor-events:0=0, 1=0, 2=0 for sensor 1.
10:24:37,688 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Initialize sensorStreamSlice end value to look for the next updated StreamCut: examples11/sensor-events:0=0, 1=0, 2=0 for sensor 0.
```
That is, as the existing version of the application relies in Flink checkpoints to periodically updated the `ReaderGroup` positions, the `StreamBookmarker` job only gets default `StreamCut`s despite events are being read. This suggests that Flink Checkpoints are not completing, or at least not updating the `ReaderGroup` state.

In a test with the new version of the app, we also observe a different side of this problem:

```
09:33:45,197 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Start bookmarking a stream slice at StreamCut: examples7/sensor-events:0=0, 1=0, 2=0 for sensor 2.
09:33:50,754 WARN  io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl    - Exception while reading from Segment : examples7/sensor-events/0.#epoch.0
io.pravega.client.stream.impl.ConnectionClosedException
        at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl.close(AsyncSegmentInputStreamImpl.java:149)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.close(SegmentInputStreamImpl.java:237)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.releaseSegmentsIfNeeded(EventStreamReaderImpl.java:191)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.updateGroupStateIfNeeded(EventStreamReaderImpl.java:160)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:92)
        at io.pravega.connectors.flink.FlinkPravegaReader.run(FlinkPravegaReader.java:188)
        at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:87)
        at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:56)
        at org.apache.flink.streaming.runtime.tasks.SourceStreamTask.run(SourceStreamTask.java:99)
        at org.apache.flink.streaming.runtime.tasks.StoppableSourceStreamTask.run(StoppableSourceStreamTask.java:45)
        at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:300)
        at org.apache.flink.runtime.taskmanager.Task.run(Task.java:711)
        at java.lang.Thread.run(Thread.java:745)
09:33:55,129 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Start bookmarking a stream slice at StreamCut: examples7/sensor-events:0=62700, 1=61275, 2=62985 for sensor 1.
09:34:05,585 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - Start bookmarking a stream slice at StreamCut: examples7/sensor-events:0=86925, 1=83790, 2=85500 for sensor 0.
09:34:21,343 WARN  io.pravega.example.flink.streamcuts.process.Bookmarker        - End bookmarking a stream slice at StreamCut: examples7/sensor-events:0=124830, 1=129105, 2=123405 for sensor 2. The slice should contain all events < 0 for a specific sensor sine wave.
09:34:21,723 WARN  io.pravega.client.stream.impl.ReaderGroupImpl                 - maxOutstandingCheckpointRequest: 3, currentOutstandingCheckpointRequest: 3, errorMessage: rejecting checkpoint request since pending checkpoint reaches max allowed limit 3
09:34:21,724 WARN  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Failed to trigger checkpoint 10 for job 3ce1e34983daddc2a2955419501a66b3. (1 consecutive failed attempts so far)
org.apache.flink.util.FlinkException: Checkpoint master hook '-1448232284' produced an exception
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.triggerHook(MasterHooks.java:176)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.triggerMasterHooks(MasterHooks.java:127)
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.triggerCheckpoint(CheckpointCoordinator.java:613)
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.triggerCheckpoint(CheckpointCoordinator.java:394)
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinator$ScheduledTrigger.run(CheckpointCoordinator.java:1235)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: io.pravega.client.stream.impl.MaxNumberOfCheckpointsExceededException: rejecting checkpoint request since pending checkpoint reaches max allowed limit
        at io.pravega.client.stream.impl.ReaderGroupImpl.initiateCheckpoint(ReaderGroupImpl.java:134)
        at io.pravega.connectors.flink.ReaderCheckpointHook.triggerCheckpoint(ReaderCheckpointHook.java:87)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook.lambda$triggerCheckpoint$0(MasterHooks.java:374)
        at org.apache.flink.util.LambdaUtil.withContextClassLoader(LambdaUtil.java:107)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook.triggerCheckpoint(MasterHooks.java:372)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.triggerHook(MasterHooks.java:153)
        ... 11 more
09:34:25,723 WARN  io.pravega.client.stream.impl.ReaderGroupImpl                 - maxOutstandingCheckpointRequest: 3, currentOutstandingCheckpointRequest: 3, errorMessage: rejecting checkpoint request since pending checkpoint reaches max allowed limit 3
09:34:25,741 WARN  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Failed to trigger checkpoint 11 for job 3ce1e34983daddc2a2955419501a66b3. (2 consecutive failed attempts so far)
org.apache.flink.util.FlinkException: Checkpoint master hook '-1448232284' produced an exception
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.triggerHook(MasterHooks.java:176)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.triggerMasterHooks(MasterHooks.java:127)
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.triggerCheckpoint(CheckpointCoordinator.java:613)
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.triggerCheckpoint(CheckpointCoordinator.java:394)
        at org.apache.flink.runtime.checkpoint.CheckpointCoordinator$ScheduledTrigger.run(CheckpointCoordinator.java:1235)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: io.pravega.client.stream.impl.MaxNumberOfCheckpointsExceededException: rejecting checkpoint request since pending checkpoint reaches max allowed limit
        at io.pravega.client.stream.impl.ReaderGroupImpl.initiateCheckpoint(ReaderGroupImpl.java:134)
        at io.pravega.connectors.flink.ReaderCheckpointHook.triggerCheckpoint(ReaderCheckpointHook.java:87)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook.lambda$triggerCheckpoint$0(MasterHooks.java:374)
        at org.apache.flink.util.LambdaUtil.withContextClassLoader(LambdaUtil.java:107)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook.triggerCheckpoint(MasterHooks.java:372)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.triggerHook(MasterHooks.java:153)
        ... 11 more
09:34:29,723 WARN  io.pravega.client.stream.impl.ReaderGroupImpl                 - maxOutstandingCheckpointRequest: 3, currentOutstandingCheckpointRequest: 3, errorMessage: rejecting checkpoint request since pending checkpoint reaches max allowed limit 3
09:34:29,734 WARN  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Failed to trigger checkpoint 12 for job 3ce1e34983daddc2a2955419501a66b3. (3 consecutive failed attempts so far)
````

The log suggests that initially Flink checkpoints are updating the `ReaderGroup` reading positions as visible in the messages containing `Start bookmarking a stream slice at StreamCut`. Once we get to the first `End bookmarking a stream slice at StreamCut`, we invoke to `generateStreamCuts()`. From that point on, we start observing Flink `CheckpointCoordinator` failures. Soon after these messages appear, the application gets blocked. Both kind of failures seems to be different manifestations of the same problem.

PS: Note that I executed the existing and new version of the app against Pravega `0.4 (master)`. While the former worked, the latter didn't. This suggests that the issue may not be related to the Server side.

**Problem location**
Pravega client/Flink connector

**Suggestions for an improvement**
Check the recent changes in Pravega client/Flink connector that may be impacted Flink Checkpoint interactions.

**Comment by shrids** [comment]
The checkpointing failed as the number of pending checkpoint reached the max allowed limit.
```
Caused by: io.pravega.client.stream.impl.MaxNumberOfCheckpointsExceededException: rejecting checkpoint request since pending checkpoint reaches max allowed limit
        at io.pravega.client.stream.impl.ReaderGroupImpl.initiateCheckpoint(ReaderGroupImpl.java:134)
        at io.pravega.connectors.flink.ReaderCheckpointHook.triggerCheckpoint(ReaderCheckpointHook.java:87)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook.lambda$triggerCheckpoint$0(MasterHooks.java:374)
        at org.apache.flink.util.LambdaUtil.withContextClassLoader(LambdaUtil.java:107)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook.triggerCheckpoint(MasterHooks.java:372)
        at org.apache.flink.runtime.checkpoint.hooks.MasterHooks.triggerHook(MasterHooks.java:153)
```

Since `generateStreamCuts()`uses the internal checkpointing mechanism the number of `generateStreamCuts()` invocations should not affect the max pending checkpoint count. I have created PR #3079  to fix this.

**Comment by RaulGracia** [comment]
Closed via #3079 (plus a change in https://github.com/pravega/pravega-samples/pull/161).
